eclipse-theia/theia
# Add vscode API `DebugSessionOptions`

#### What it does
Closes #8518 

Adds the `vscode.debug.DebugSessionOptions` API to Theia. This enables the following features:

- Show child debug sessions as a tree under the parent sessions (Theia used to show them as siblings)
    - The new `compact` property will simplify the label and tree structure if possible
- Different debug sessions no longer share the same `DebugConsoleSession` and therefore are displayed independently
    - Child debug sessions with the `consoleMode` property set to `MergeWithParent` will continue to share their console session with each other.
    - You can switch between console sessions with a selector, assuming more than one session is currently running.

The changes to `DebugConsoleSession` are slightly breaking. While it's still possible to simply inject the `DebugConsoleSession` directly, its tight coupling to `DebugSession` makes that impractical. Instead of injecting the `DebugConsoleSession` directly, a `DebugConsoleSessionFactory` allows to create a `DebugConsoleSession` with a provided `DebugSession`.

#### How to test


1. Clone [this test repo](https://github.com/msujew/debug-session-options-example) into Theia and make sure the `vscode-js-debug` extension is installed in Theia.
2. Start a debugging session with the contained `Launch Program` configuration.
3. Child processes will now be shown as a tree in the threads view.
4. Start a second debugging session with the same configuration
5. Another tree will appear in the threads view and you can switch between the different debug consoles.
6. You can add `configuration.consoleMode = undefined;` and/or `configuration.compact = undefined;` after [this line](https://github.com/msujew/theia/blob/e26d3bff923e51f013aece29ed7a9db8eb27efce/packages/plugin-ext/src/main/browser/debug/debug-main.ts#L273), and restart the server to test how these properties influence the debug console/view.

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
